pubnub/lib/src/crypto/crypto.dart

```diff
@@ -1,132 +1,113 @@
-import 'encryption_mode.dart';
-
-/// Configuration used in cryptography.
-class CryptoConfiguration {
-  /// Encryption mode used.
-  final EncryptionMode encryptionMode;
+import 'package:pubnub/src/crypto/aesCbcCryptor.dart';
+import 'package:pubnub/src/crypto/legacyCryptor.dart';
+import 'cryptoConfiguration.dart';
+import 'cryptorHeader.dart';
 
-  /// Whether key should be encrypted.
-  final bool encryptKey;
+/// CryptoModule is responsible for encryption and decryption
+/// of PubNub messages.
+class CryptoModule implements ICryptoModule {
+  final ICryptor defaultCryptor;
+  List<ICryptor>? cryptors;
+
+  late CryptoConfiguration defaultConfiguration;
+  late LegacyCryptoModule legacyCryptoModule;
+
+  CryptoModule(
+      {required this.defaultCryptor,
+      this.cryptors,
+      this.defaultConfiguration = const CryptoConfiguration()}) {
+    legacyCryptoModule =
+        LegacyCryptoModule(defaultConfiguration: defaultConfiguration);
+  }
 
-  /// Whether a random IV should be used.
-  final bool useRandomInitializationVector;
+  factory CryptoModule.legacyCryptoModule(CipherKey cipherKey,
+      {defaultCryptoConfiguration = const CryptoConfiguration()}) {
+    return CryptoModule(
+        defaultCryptor: LegacyCryptor(cipherKey,
+            cryptoConfiguration: defaultCryptoConfiguration),
+        cryptors: <ICryptor>[AesCbcCryptor(cipherKey)]);
+  }
 
-  const CryptoConfiguration(
-      {this.encryptionMode = EncryptionMode.CBC,
-      this.encryptKey = true,
-      this.useRandomInitializationVector = true});
-}
+  factory CryptoModule.aesCbcCryptoModule(CipherKey cipherKey,
+      {defaultCryptoConfiguration = const CryptoConfiguration()}) {
+    return CryptoModule(
+        defaultCryptor: AesCbcCryptor(cipherKey),
+        cryptors: <ICryptor>[
+          LegacyCryptor(cipherKey,
+              cryptoConfiguration: defaultCryptoConfiguration),
+        ]);
+  }
 
-/// Default cryptography module used in PubNub SDK.
-class CryptoModule implements ICryptoModule {
-  final CryptoConfiguration defaultConfiguration;
-
-  /// Default configuration is:
-  /// * `encryptionMode` set to [EncryptionMode.CBC].
-  /// * `encryptKey` set to `true`.
-  /// * `useRandomInitializationVector` set to `true`.
-  CryptoModule({this.defaultConfiguration = const CryptoConfiguration()});
-
-  crypto.Key _getKey(CipherKey cipherKey, CryptoConfiguration configuration) {
-    if (configuration.encryptKey) {
-      return crypto.Key.fromUtf8(
-          sha256.convert(cipherKey.data).toString().substring(0, 32));
-    } else {
-      return crypto.Key(Uint8List.fromList(cipherKey.data));
-    }
+  @override
+  List<int> encrypt(List<int> data) {
+    _emptyContentValidation(data);
+    var encrypted = defaultCryptor.encrypt(data);
+    if (encrypted.metadata.isEmpty) return encrypted.data;
+
+    var header =
+        CryptorHeader.from(defaultCryptor.identifier, encrypted.metadata);
+    var headerData = List<int>.filled(header!.length, 0);
```

Comment:
Are you sure that `header` is not null here?

Reply:
So, This is encryption code. Before attempting CryptoHeader.from(...), Here is what we did:
```
    var encrypted = defaultCryptor.encrypt(data);
    if (encrypted.metadata.isEmpty) return encrypted.data;
```

Which means If the default cryptor is not giving metadata(that means legacy one) we return from there only.

_**If we haven't return that means It's not legacy cryptor**_ and If it's not legacy cryptor then we must have header.[going forward we won't ever support cryptor which does not conforms our header requirement]  So header! will never throw.

I think I know the root of confusion:
from(..) has this `if (id == LEGACY_IDENTIFIER) return null;`
This was there for testing purpose(If CryptoHeader.from() called with legacy id then it should work as expected),  while consuming this method in encrypt method. We are sure that execution will reach to Header.from(...) only when we have non-legacy cryptor.


I think I can remove `if (id == LEGACY_IDENTIFIER) return null;` condition. Because from method will be called only when we deal with non legacy cryptor.